File: src/multi_tf/scripts/multi_subscriber.py
#!/usr/bin/env python3

"""  
    Implementation steps:   
        1.Import module
        2.Initialize the ros node
        3.Create a TF subscription object
        4.Figure out the coordinates of son1 with respect to son2
        5.Create a coordinate point dependent on son1 and call the API to figure out the coordinates of that point in son2
        6.spin
"""

# 1.Import module
import rospy
import tf2_ros
from geometry_msgs.msg import TransformStamped
from tf2_geometry_msgs import PointStamped

if __name__ == "__main__":

    # 2.Initialize the ros node
    rospy.init_node("frames_sub_p")
    # 3.Create a TF subscription object
    buffer = tf2_ros.Buffer()
    listener = tf2_ros.TransformListener(buffer)

    rate = rospy.Rate(1)
    while not rospy.is_shutdown():
        try:
            # 4.Figure out the coordinates of son1 with respect to son2
            # lookup_transform(self, target_frame, source_frame, time, timeout=rospy.Duration(0.0)):
            tfs = buffer.lookup_transform("son2","son1",rospy.Time(0))
            rospy.loginfo("Relative relation between son1 and son2:")
            rospy.loginfo("Parent coordinate system: %s", tfs.header.frame_id)
            rospy.loginfo("Child coordinate system: %s", tfs.child_frame_id)
            rospy.loginfo("Relative coordinates: x = %.2f, y = %.2f, z = %.2f",
                        tfs.transform.translation.x,
                        tfs.transform.translation.y,
                        tfs.transform.translation.z,)
            
            # 5.Create a coordinate point dependent on son1 and call the API to figure out the coordinates of that point in son2
            point_source = PointStamped()
            point_source.header.frame_id = "son1"
            point_source.header.stamp = rospy.Time.now()
            point_source.point.x = 1
            point_source.point.y = 1
            point_source.point.z = 1

            point_target = buffer.transform(point_source,"son2",rospy.Duration(0.5))

            rospy.loginfo("point_target Coordinate system: %s", point_target.header.frame_id)
            rospy.loginfo("The coordinates with respect to son2: (%.2f, %.2f, %.2f)",
                        point_target.point.x,
                        point_target.point.y,
                        point_target.point.z)

        except Exception as e:
            rospy.logerr("Error: %s", e)

        rate.sleep()
    # 6.spin    
    # rospy.spin()

File: src/multi_tf/src/multi_subscriber.cpp
/*
    Implementation steps:
        1.Include header file
        2.Initialize the ros node
        3.Create the ros handle
        4.Create a TF subscription object
        5.Obtain the coordinates of the origin of son1 coordinate system in son2 from the parsing subscription information
        Resolve the coordinates of points in son1 with respect to son2
        6.spin
*/

// 1.Include header file
#include "ros/ros.h"
#include "tf2_ros/transform_listener.h"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.h"
#include "geometry_msgs/TransformStamped.h"
#include "geometry_msgs/PointStamped.h"

int main(int argc, char *argv[])
{   
    // 2.Initialize the ros node
    ros::init(argc,argv,"sub_frames");
    // 3.Create the ros handle
    ros::NodeHandle nh;
    // 4.Create a TF subscription object
    tf2_ros::Buffer buffer; 
    tf2_ros::TransformListener listener(buffer);
    // 5.Obtain the coordinates of the origin of son1 coordinate system in son2
    ros::Rate r(1);
    while (ros::ok())
    {
        try
        {
            // Resolve the coordinates of points in son1 with respect to son2
            geometry_msgs::TransformStamped tfs = buffer.lookupTransform("son2","son1",ros::Time(0));
            ROS_INFO("Son1 with respect to Son2: Parent coordinate system ID = %s", tfs.header.frame_id.c_str());
            ROS_INFO("Son1 with respect to Son2: Child coordinate system ID = %s", tfs.child_frame_id.c_str());
            ROS_INFO("Son1 with respect to Son2: x = %.2f, y = %.2f, z = %.2f",
                    tfs.transform.translation.x,
                    tfs.transform.translation.y,
                    tfs.transform.translation.z);

            // Coordinate point resolution
            geometry_msgs::PointStamped ps;
            ps.header.frame_id = "son1";
            ps.header.stamp = ros::Time::now();
            ps.point.x = 1.0;
            ps.point.y = 2.0;
            ps.point.z = 3.0;

            geometry_msgs::PointStamped psAtSon2;
            psAtSon2 = buffer.transform(ps, "son2");
            ROS_INFO("Coordinates in Son2: x = %.2f, y = %.2f, z = %.2f",
                    psAtSon2.point.x,
                    psAtSon2.point.y,
                    psAtSon2.point.z);
        }
        catch(const std::exception& e)
        {
            ROS_INFO("Error: %s", e.what());
        }

        r.sleep();
        // 6.spin
        ros::spinOnce();
    }
    return 0;
}